File: 2021/go/day12/part1/part1.go
package part1

import (
	"bufio"
	"strconv"
	"strings"
)

func Solve(input string) string {
	scanner := bufio.NewScanner(strings.NewReader(input))
	var lines []string
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return strconv.Itoa(NewGraph(lines).DFS())
}

type Graph struct {
	edges map[string][]string
	paths int
}

func NewGraph(lines []string) *Graph {
	edges := make(map[string][]string)
	for _, line := range lines {
		s := strings.Split(line, "-")
		n1, n2 := s[0], s[1]
		edges[n1] = append(edges[n1], n2)
		edges[n2] = append(edges[n2], n1)
	}
	return &Graph{edges: edges}
}

func (g *Graph) DFS() int {
	g.paths = 0
	g.dfs("start", map[string]bool{})
	return g.paths
}

func (g *Graph) dfs(node string, visited map[string]bool) {
	if node == "end" {
		g.paths += 1
		return
	}

	visited[node] = true
	defer func() { visited[node] = false }()

	for _, next := range g.edges[node] {
		if !visited[next] || canVisitMultiTimes(next) {
			g.dfs(next, visited)
		}
	}
}

func canVisitMultiTimes(node string) bool {
	return strings.ToUpper(node) == node
}

File: 2021/elixir/apps/day12/lib/part1.ex
defmodule Day12Part1 do
  @moduledoc """
  Day12 - Part1
  """

  def solve(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.map(&String.split(&1, "-"))
    |> Enum.reduce(%{}, fn [s | [e]], map ->
      map
      |> Map.update(s, [e], fn edges -> [e | edges] end)
      |> Map.update(e, [s], fn edges -> [s | edges] end)
    end)
    |> dfs()
    |> length()
  end

  defp dfs(map) do
    dfs("start", map, MapSet.new(), [], [])
  end

  defp dfs(node, _map, _visited, path, paths) when node == "end" do
    [
      Enum.reverse(["end" | path]) | paths
    ]
  end

  defp dfs(node, map, visited, path, paths) do
    if map[node] == nil or node in visited do
      # no path from here
      paths
    else
      visited = if(multi_visitable?(node), do: visited, else: MapSet.put(visited, node))

      Enum.reduce(map[node], paths, fn next, paths ->
        dfs(next, map, visited, [node | path], paths)
      end)
    end
  end

  defp multi_visitable?(node) do
    String.upcase(node) == node
  end
end
